In [1]:
import numpy as np 
import matplotlib.pyplot as plt 
%matplotlib inline 
import pandas as pd 
import json

### Загрузка данных

In [2]:
data = pd.read_csv('~/comptech/420_redirects.csv', index_col=0)

In [3]:
data.head()

redir_id  rule_id  start_date                            query  \
0     10303    23358  2020-08-19                   dior хайлайтер   
1     10304    23359  2020-08-19                    dior хайлатер   
2     10319    23374  2020-08-19  diorskin mineral nude luminizer   
3     10476    23531  2020-08-19                   хайдайтер dior   
4     10477    23532  2020-08-19                   хайлайтер dior   

   category_id  parent_id    category  
0      3327621  3327595.0  Хайлайтеры  
1      3327621  3327595.0  Хайлайтеры  
2      3327621  3327595.0  Хайлайтеры  
3      3327621  3327595.0  Хайлайтеры  
4      3327621  3327595.0  Хайлайтеры

In [4]:
categories = pd.read_csv('~/comptech/categories.csv', index_col=0)

In [5]:
categories.head()

id  parent_id  site_id  external_id                      name
0  3327511  3328184.0      420         1471                       Cow
1  3327512  3328184.0      420         1470                    Trolls
2  3327516  3328184.0      420         1466                 Фельдшер 
3  3658304  3327503.0      420         3641         Наборы одеколонов
4  3658305  3328459.0      420         3642  Искусство дарить подарки

In [6]:
products = pd.read_csv('~/comptech/products.csv', index_col=0)

In [7]:
products.head()

id  site_id  external_id  \
0  2573904266      420  24944900083   
1  2573904322      420  24944900078   
2  2573904375      420   7005200014   
3  2573904490      420  24944900057   
4  2573905005      420  17059600013   

                                                name        brand  \
0  Губная помада  | URBAN DECAY | VICE LIPSTICK M...  URBAN DECAY   
1  Губная помада  | URBAN DECAY | VICE LIPSTICK M...  URBAN DECAY   
2    Губная помада | BOURJOIS | Rouge Edition Velvet     BOURJOIS   
3  Губная помада  | URBAN DECAY | VICE LIPSTICK C...  URBAN DECAY   
4     Блеск для губ  | GA DE | VELVETEEN ULTRA SHINE        GA DE   

                                         description last_existed  group_id  
0  Губная помада Urban Decay Vice Lipstick незаме...   2021-01-28    111948  
1  Губная помада Urban Decay Vice Lipstick - это ...   2021-01-28    111946  
2  Бархатный флюид для губ Bourjois Rouge edition...   2021-01-28    113391  
3  Губная помада Urban Decay Vice Lipstick рассчи...   2021-01-28    123428  
4  Оригинальная формула объединяет комплекс пигме...   2021-01-28     93766

In [8]:
products_categories = pd.read_csv('~/comptech/products_categories.csv', index_col=0)

In [9]:
products_categories.head()

site_id  product_id  category_id
0      420  2573916522       873202
1      420  2574226585       873224
2      420  2585316813       873244
3      420  2573916527       873202
4      420  2612144294      3115485

In [10]:
#with open ('sessions_420.json') as f:
#    tmp = json.load(f)

In [11]:
df = pd.read_csv('~/comptech/320-redirect.tsv',sep='\t')

In [12]:
df.head()

query  count     first_category    second_category  url  \
0  samsung 7170    474          Смартфоны          Смартфоны  NaN   
1     realme 3x     16          Смартфоны          Смартфоны  NaN   
2     whirlpool   3875  Стиральные машины  Стиральные машины  NaN   
3    фритюрница  11641         Фритюрницы        Мультиварки  NaN   
4   samsung s10   5388          Смартфоны          Смартфоны  NaN   

   viewed_category_1  view_count_1  viewed_category_2  view_count_2  \
0         Телевизоры           166         Телевизоры           166   
1          Смартфоны             9          Смартфоны             9   
2  Стиральные машины           221  Стиральные машины           221   
3         Фритюрницы          6357           Пылесосы            26   
4          Смартфоны          1974          Смартфоны          1968   

     viewed_category_3  view_count_3  redirect_proba  
0            Смартфоны             7        0.000108  
1                 Фены             2        0.000015  
2         Холодильники           205        0.000963  
3         Электрогрили            26        0.938724  
4  Чехлы для телефонов           114        0.000979

### Построения дерева категорий

In [13]:
categories.head()

id  parent_id  site_id  external_id                      name
0  3327511  3328184.0      420         1471                       Cow
1  3327512  3328184.0      420         1470                    Trolls
2  3327516  3328184.0      420         1466                 Фельдшер 
3  3658304  3327503.0      420         3641         Наборы одеколонов
4  3658305  3328459.0      420         3642  Искусство дарить подарки

In [14]:
categories.nunique()

id             2483
parent_id       285
site_id           1
external_id    2483
name           1881
dtype: int64

In [15]:
categories.shape

(2483, 5)

In [16]:
products_categories.head()

site_id  product_id  category_id
0      420  2573916522       873202
1      420  2574226585       873224
2      420  2585316813       873244
3      420  2573916527       873202
4      420  2612144294      3115485

In [17]:
products_categories.shape

(112858, 3)

In [18]:
products_categories.isnull().sum()

site_id        0
product_id     0
category_id    0
dtype: int64

In [19]:
products_categories.nunique()

site_id            1
product_id     57925
category_id      361
dtype: int64

In [20]:
products_by_categories = products_categories.groupby('category_id')['product_id'].apply(list)

In [21]:
products_by_categories

category_id
873100     [2575311232, 2594251213, 2592281923, 259228192...
873153     [2611053879, 2611053878, 2611053902, 258086265...
873170     [2573917020, 2573917060, 2573909828, 257391705...
873172     [2573916531, 2573916532, 2573916506, 257391650...
873174     [2598111436, 2596859705, 2596859706, 259685970...
                                 ...                        
3658303    [2573917109, 2573917083, 2573916681, 257391728...
3671663    [2573913520, 2573919584, 2614212973, 257677169...
3672618    [2587793982, 2587793981, 2573916482, 261494262...
3709154                                         [2573914475]
3721283    [2578496883, 2611053920, 2584864564, 261482345...
Name: product_id, Length: 361, dtype: object

In [22]:
len(products_by_categories)

361

То есть из общего числа категорий 2483 только в 361 из них есть продукты.

In [23]:
categories.isna().sum()

id             0
parent_id      2
site_id        0
external_id    0
name           0
dtype: int64

In [24]:
categories.dropna(inplace=True)

In [25]:
categories.isna().sum()

id             0
parent_id      0
site_id        0
external_id    0
name           0
dtype: int64

In [26]:
categories.parent_id = categories.parent_id.astype('int64')

In [27]:
categories.parent_id.nunique()

285

In [28]:
categories.head()

id  parent_id  site_id  external_id                      name
0  3327511    3328184      420         1471                       Cow
1  3327512    3328184      420         1470                    Trolls
2  3327516    3328184      420         1466                 Фельдшер 
3  3658304    3327503      420         3641         Наборы одеколонов
4  3658305    3328459      420         3642  Искусство дарить подарки

In [29]:
categories_by_parent = categories.groupby('parent_id')['id'].apply(list)

In [30]:
categories_by_parent

parent_id
872901     [3345097, 3597453, 3658308, 3658309, 3658310, ...
872904     [2825483, 873819, 873820, 873821, 873822, 8738...
872907     [2825482, 873847, 873848, 873850, 2796250, 873...
872912     [2825481, 873866, 873868, 873870, 873871, 8738...
872915     [2825480, 3139390, 873893, 873894, 873904, 873...
                                 ...                        
3559134        [3658327, 3559130, 3559131, 3559132, 3559135]
3559138                          [3559139, 3559142, 3559143]
3658299                                   [3658301, 3658300]
3658305                                            [3658306]
3709153                                            [3709154]
Name: id, Length: 285, dtype: object

In [31]:
products_by_categories

category_id
873100     [2575311232, 2594251213, 2592281923, 259228192...
873153     [2611053879, 2611053878, 2611053902, 258086265...
873170     [2573917020, 2573917060, 2573909828, 257391705...
873172     [2573916531, 2573916532, 2573916506, 257391650...
873174     [2598111436, 2596859705, 2596859706, 259685970...
                                 ...                        
3658303    [2573917109, 2573917083, 2573916681, 257391728...
3671663    [2573913520, 2573919584, 2614212973, 257677169...
3672618    [2587793982, 2587793981, 2573916482, 261494262...
3709154                                         [2573914475]
3721283    [2578496883, 2611053920, 2584864564, 261482345...
Name: product_id, Length: 361, dtype: object

In [32]:
selected_categories = products_by_categories.index.tolist()

In [33]:
len(selected_categories)

361

In [34]:
parent_list = categories_by_parent.index.tolist()

In [35]:
len(parent_list)

285

In [36]:
#словарь: ключи - родительские категории
category_tree = dict.fromkeys(parent_list)

In [37]:
for name in category_tree.keys():
    category_tree[name] = dict.fromkeys(categories_by_parent[name])
    for subname in category_tree[name].keys():
        if subname in selected_categories:
            category_tree[name][subname] = products_by_categories[subname]

Структура дерева: list_of_products = category_tree[parent][category]

In [38]:
category_tree

{872901: {3345097: [2595151172,
   2595151170,
   2595151198,
   2595151196,
   2595151197,
   2595151194,
   2583089945,
   2583048695,
   2582976613,
   2582976612,
   2582976615,
   2582976611,
   2582976610,
   2582976617,
   2582976616,
   2582976619,
   2582976618,
   2582976597,
   2582976596,
   2582976599,
   2582976598,
   2582976593,
   2582976592,
   2582976595,
   2582976594,
   2582976600,
   2582976589,
   2582976588,
   2582976591,
   2582976590,
   2582976587,
   2582976586],
  3597453: [2573912357,
   2573923439,
   2573935199,
   2573909862,
   2573909863,
   2573906452,
   2573906453,
   2574051323,
   2574051262,
   2573931275,
   2573932056,
   2573916527,
   2573915366,
   2573914052,
   2573924455,
   2581003116,
   2573929966,
   2573929955,
   2581004437,
   2581004426,
   2580363078,
   2573938298,
   2612977210,
   2612977208,
   2573905913,
   2573905910,
   2573905520,
   2573905450,
   2573905440,
   2573905447,
   2573905444,
   2573905458,
   2573921645

In [39]:
products.head()

id  site_id  external_id  \
0  2573904266      420  24944900083   
1  2573904322      420  24944900078   
2  2573904375      420   7005200014   
3  2573904490      420  24944900057   
4  2573905005      420  17059600013   

                                                name        brand  \
0  Губная помада  | URBAN DECAY | VICE LIPSTICK M...  URBAN DECAY   
1  Губная помада  | URBAN DECAY | VICE LIPSTICK M...  URBAN DECAY   
2    Губная помада | BOURJOIS | Rouge Edition Velvet     BOURJOIS   
3  Губная помада  | URBAN DECAY | VICE LIPSTICK C...  URBAN DECAY   
4     Блеск для губ  | GA DE | VELVETEEN ULTRA SHINE        GA DE   

                                         description last_existed  group_id  
0  Губная помада Urban Decay Vice Lipstick незаме...   2021-01-28    111948  
1  Губная помада Urban Decay Vice Lipstick - это ...   2021-01-28    111946  
2  Бархатный флюид для губ Bourjois Rouge edition...   2021-01-28    113391  
3  Губная помада Urban Decay Vice Lipstick рассчи...   2021-01-28    123428  
4  Оригинальная формула объединяет комплекс пигме...   2021-01-28     93766

In [40]:
products.nunique()

id              60291
site_id             1
external_id     60291
name            34963
brand             989
description     32087
last_existed       13
group_id        34435
dtype: int64

In [41]:
products.isnull().sum()

id                0
site_id           0
external_id       0
name              0
brand           198
description       0
last_existed      0
group_id          0
dtype: int64